src/pending.js

```diff
@@ -0,0 +1,128 @@
+// thresholds for pending tasks.
+const MIN_CHECK_AGE = 300; // seconds
+
+// calculate the taskKey and scheduled date with the earliest scheduled time for
+// the given workerType; if there are none, returns null and the current time
+var earliest = (state, workerType, now) => {
+  return _.reduce(state.pendingTasks[workerType], (res, scheduled, taskKey) => {
+    if (scheduled < res.scheduled) {
+      return {scheduled, taskKey};
+    } else {
+      return res;
+    }
+  }, {taskKey: null, scheduled: now});
+};
+
+// update the state based on a task message
+var update = (state, workerType, taskKey, isPending, scheduled) => {
+  let workerTypeState = state.pendingTasks[workerType];
+
+  if (!workerTypeState) {
+    state.pendingTasks[workerType] = workerTypeState = {};
+  }
+
+  if (isPending) {
+    debug('task pending: %s at %s (%s)', taskKey, scheduled, workerType);
+    workerTypeState[taskKey] = new Date(scheduled).getTime();
+  } else {
+    debug('task no longer pending: %s (%s)', taskKey, workerType);
+    if (workerTypeState[taskKey]) {
+      state.readyWorkerTypes[workerType] = true;
+      delete workerTypeState[taskKey];
+    }
+  }
+};
+
+// update monitors based on the current state
+var flush = (state, monitor, now) => {
+  for (let workerType of Object.keys(state.readyWorkerTypes)) {
+    let {taskKey, scheduled} = earliest(state, workerType, now);
+    let waiting = taskKey ? now - scheduled : 0;
+    monitor.measure(`tasks.${workerType}.pending`, waiting);
+  }
+};
+
+// For each worker type with a task we think is pending for over MIN_CHECK_AGE,
+// poll the status of that task and, if it is actually not pending, remove it
+// from the list.  This provides a way to catch cases where the service misses
+// a pulse message or they are delivered out of order.
+var check = (state, queue, now) => {
+  return Promise.all(Object.keys(state.pendingTasks).map(async (workerType) => {
+    let {taskKey, scheduled} = earliest(state, workerType, now);
+    if (taskKey && now - scheduled > MIN_CHECK_AGE * 1000) {
+      let [taskId, runId] = taskKey.split('/');
+      let taskStatus = await queue.status(taskId);
+      if (taskStatus.status.runs[runId].state !== 'pending') {
+        debug('task not actually pending: %s (%s)', taskKey, workerType);
+        delete state.pendingTasks[workerType][taskKey];
+      }
+    }
+  }));
+};
+
+module.exports = ({monitor, listener, credentials}) => {
+  let state = {
+    // mappings from task key to pending time, keyed by workerType; an empty list
+    // here is significant in that it means there is nothing pending for that
+    // workerType
+    pendingTasks: {},
+
+    // workerTypes for which we have seen a single task both enter and exit the
+    // pending state, and thus have a reasonable expectation of knowing the
+    // total pending time for, even if the stats-collector process has just
+    // started.
+    readyWorkerTypes: {},
+  };
```

Comment:
Like  I said above I suggest refactoring this into a single flat list... or maybe two... but a list per workerType.

Also consider making a class and let this be properties on the class... Will probably mean less passing state around.